File: DESIGN-CSharp/AbstractFactory/Adventurer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AbstractFactoy
{
    public abstract class Adventurer
    {
        protected internal Weapon weapon;
        
        protected internal Clothes clothes;

        public abstract void display();
    }

    public class Archer : Adventurer
    {
        public Archer()
        {
            EquipmentFactory archerEquipmentFactory = new ArcherEquipmentFactory();
            weapon = archerEquipmentFactory.produceWeapon();
            clothes = archerEquipmentFactory.produceClothes();
        }

        public override void display()
        {
            Console.WriteLine("this is an archer; show equipment data");
            weapon.display();
            clothes.display();
            Console.WriteLine("-------------------------------------");
        }
    }

    public class Warrior : Adventurer
    {
        public Warrior()
        {
            EquipmentFactory warriorEquipmentFactory = new WarriorEquipmentFactory();
            weapon = warriorEquipmentFactory.produceWeapon();
            clothes = warriorEquipmentFactory.produceClothes();
        }

        public override void display()
        {
            Console.WriteLine("this is a warrior; show equipment data");
            weapon.display();
            clothes.display();
            Console.WriteLine("-------------------------------------");
        }
    }
}

File: CreationalPatterns/AbstractFactory/Adventurer.py
from EquipmentFactory import WarriorEquipmentFactory, ArcherEquipmentFactory

class Adventurer:
    def __init__(self):
        self.weapon = None
        self.clothes = None

    def display(self):
        pass

class Archer(Adventurer):
    def __init__(self):
        super().__init__() #super().__init__() 來調用父類的構造函數
        archerEquipmentFactory = ArcherEquipmentFactory()
        self.weapon = archerEquipmentFactory.produceWeapon()
        self.clothes = archerEquipmentFactory.produceClothes()

    def display(self):
        print("this is an archer; show equipment data")
        self.weapon.display()
        self.clothes.display()
        print("-------------------------------------")

class Warrior(Adventurer):
    def __init__(self):
        super().__init__()
        warriorEquipmentFactory = WarriorEquipmentFactory()
        self.weapon = warriorEquipmentFactory.produceWeapon()
        self.clothes = warriorEquipmentFactory.produceClothes()

    def display(self):
        print("this is a warrior; show equipment data")
        self.weapon.display()
        self.clothes.display()
        print("-------------------------------------")